### PR TITLE
Validating content type ability

### PR DESCRIPTION
**Pull request by asiniy**
Todo:
- [x] Make it working
- [x] Rename `mime_type` to `content_type`
- [x] Create specific processor and wrap file metadata to it
- [x] Create common file types, like: `video`, `image` and so on
- [x] Change default configs to preprocess metadata for file (or find another way to preprocess on validate)
- [x] Refactor
- [x] Docs

I'm pairing w/ @stepantubanov

This is for mime type validation purposes only. It's bad solution to trust to the file format, so we preferred to implement mime types first and then hardcode file types to extensions relation. So, we'll implement format validation next PR

@keichan34 what do you think about direction?

### CODE REVIEW

**Comment by keichan34** [comment]
Thanks! I was thinking, it would be nice to have a processor (something like the [ImageMagick metadata extractor](https://github.com/keichan34/exfile-imagemagick/blob/master/lib/exfile_imagemagick/metadata.ex)) that would first extract the mime type (and filename (would help with #22 )) and put it in the `Exfile.File`'s meta area. Then, the validation can just check the metadata to see if it's valid or not. 

The reasoning for this is mainly because the user may want to store the mime type and/or filename somewhere else, and it doesn't make sense to use the file command multiple times.

**Comment by asiniy** [comment]
@keichan34 yeah, that's better.

2 questions:
1) As I understood correctly, I have to processor before file is uploaded. How can I do that? Could you provide me code sample?
2) How could I name this module? I think, `Exfile.Metadata` is a good catch, but may conflict with existing processor for `exfile-imagemagick` processor. WDYT?

**Comment by keichan34** [comment]
1) You can set preprocessors in the backend configuration ([example here](https://github.com/keichan34/exfile/blob/master/config/test.exs#L17)). I'm still thinking if this is the best idea or not, but I think turning it on by default would be a good idea for usability -- [probably around here](https://github.com/keichan34/exfile/blob/5d10221a62f1198855076888f2fee43e18a6ca58/lib/exfile/backend.ex#L81)

2) `ExfileImagemagick.Metadata` is the name of the module in `exfile-imagemagick`, so I don't think there would be a conflict. :) Maybe `Exfile.Processor.Metadata` or `Exfile.MetadataProcessor` might make more sense?

(The basic summary of pre/post-processors are that the preprocessors get run on the file before it's uploaded, and postprocessors are run on the file after it's downloaded from the backend. I use them to encrypt files before storing them in the backend and decrypting them on demand)

**Comment by asiniy** [comment]
@keichan34 ok, I will continue tomorrow. btw, I decided to rename `mime_type` to `content_type` since `content_type` is definitely more popular term

**Comment by keichan34** [comment]
Got it. Thanks again!

**Comment by asiniy** [comment]
Hey!

We made tests passing today + started to work on processor.

Here my thoughts on preprocessing be default. Generally, I think it's not a best way because I think we should have a lot of validators, like: `FileSize`, `ContentType`, `Format` and so on. Do you want to run these commands for each model, even we have validation just in one?

I think it's better to make a lazy implementation of processor, ie for [ExfileImagemagick.Metadata](https://github.com/keichan34/exfile-imagemagick/blob/master/lib/exfile_imagemagick/metadata.ex):

``` elixir
def call(file: %Exfile.File{ meta: meta }) when meta != %{} do
  { :ok, file }
end

def call(file, _, _) do
  # previous stuff
end
```

WDYT?

**Comment by keichan34** [comment]
Hm. What about having the validation and metadata extraction steps separate? So the preprocessor will run file once, and maybe extract some of the file size metadata as well and put it in the `meta` map. Then, we can change `Exfile.Ecto.ValidateMimeType.validate_mime_type/4` to just look at the values in the `meta` map and add the validation error if necessary. If they aren't there, an error can be raised ("Please make sure you're using the metadata preprocessor", etc)

**Comment by asiniy** [comment]
@keichan34 but we'll spend some tome otherwise (if user doesn't want to validate something). I think it'll slow down app, esp. on high loads. I have an idea in my mind, let me return with it

**Comment by asiniy** [comment]
@keichan34 here is out solution for preventing second evaluation: https://github.com/keichan34/exfile/pull/28/commits/b8c92590b9e3ecfe4258533c5d9b132b780b3f09#diff-3bfe57a2d63ec90b6956a86c8a595cebR4 for not running processor twice.

We also tested it: https://github.com/keichan34/exfile/pull/28/commits/b8c92590b9e3ecfe4258533c5d9b132b780b3f09#diff-2ad79033faef1902fab8ac68e1a190f5R20

Tomorrow we apply processor to validator and will change changeset according to this processor results

**Comment by asiniy** [comment]
@keichan34 @stepantubanov ready for review

File: lib/exfile/ecto/validate_content_type.ex [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    end
+  end
+
+  defp perform_validation(changeset, field, accepted_content_types, message) do
+    { :ok, processed_file } = changeset
+      |> Changeset.get_change(field)
+      |> ContentTypeProcessor.call([], [])
+
+    changeset_with_processed_file = changeset
+      |> Changeset.put_change(field, processed_file)
```
Comment (asiniy) [inline lib/exfile/ecto/validate_content_type.ex]: I found error here!

We are getting `%Exfile.File{}` struct here, but putting `%Exfile.Localfile{}` to changeset back! This is the reason why things like https://github.com/keichan34/exfile/blob/master/lib/exfile/phoenix/helpers.ex#L76 can't pattern match

Comment (asiniy) [inline lib/exfile/ecto/validate_content_type.ex]: Same thing for #35

File: lib/exfile/processor/content_type.ex [hunk 2]
```diff
@@ -0,0 +1,25 @@
+defmodule Exfile.Processor.ContentType do
+  @behaviour Exfile.Processor
+
+  def call(file = %{ meta: %{ "content_type" => content_type }}, _, _) when is_binary(content_type) do
+    { :ok, file }
+  end
+
+  def call(file, _, _) do
+    file = Exfile.ProcessorChain.coerce_to_local_file(file)
```
Comment (asiniy) [inline lib/exfile/processor/content_type.ex]: I found error here!

We are getting `%Exfile.File{}` struct here, but putting `%Exfile.Localfile{}` to changeset back! This is the reason why things like https://github.com/keichan34/exfile/blob/master/lib/exfile/phoenix/helpers.ex#L76 can't pattern match

Same thing for #35

**Comment by keichan34** [comment]
Sorry, it's going to take a few more days for me to get around to reviewing this. Thanks for your patience!

**Comment by asiniy** [comment]
np

**Comment by asiniy** [comment]
@keichan34 I want to make new PRs, but they are blocked by these two!

**Comment by keichan34** [comment]
@asiniy Just a couple minor problems, then I should be able to merge it. Sorry it took so long! I've pushed the fixes to https://github.com/keichan34/exfile/tree/validate_file_mime_type

**Comment by asiniy** [comment]
@keichan34 what do you mean by minor problems?

**Comment by keichan34** [comment]
Sorry, here are the two fixes that I made:

70b57f6dca9a4dae35b2e422c709d14d8983cfbe
f9a8b0465c61169dd50ea067fe3d83964d1b5532

**Comment by asiniy** [comment]
Awesome, so I need to fix them in this branch, right?

**Comment by keichan34** [comment]
Yup. Just put those in to your commit and I'll be ready to merge. :)

**Comment by asiniy** [comment]
@keichan34 done

**Comment by keichan34** [comment]
Thanks 🎉 Sorry again for the delay.

**Pull request merged**